### PR TITLE
MergeAdjacentBarriers to first check if barriers are adjacent.

### PR DESCRIPTION
**Pull request by kdk**
#3273 introduced a regression in the RB benchmarks:

https://qiskit.github.io/qiskit/#randomized_benchmarking.RandomizedBenchmarkingBenchmark.time_ibmq_backend_transpile_single_thread?p-rb_pattern=%5B%5B0%5D%5D&commits=b0395c1d
https://qiskit.github.io/qiskit/#randomized_benchmarking.RandomizedBenchmarkingBenchmark.time_ibmq_backend_transpile_single_thread?p-rb_pattern=%5B%5B0%2C%201%5D%5D&commits=b0395c1d
https://qiskit.github.io/qiskit/#randomized_benchmarking.RandomizedBenchmarkingBenchmark.time_ibmq_backend_transpile?p-rb_pattern=%5B%5B0%2C%201%5D%5D&commits=b0395c1d

Profiling locally, a substantial amount of time was added in `MergeAdjacentBarriers._collect_potential_merges` which is run automatically through `BarrierBeforeFinalMeasurements`.  The existing test of whether or not two barriers can be merged while maintaining the validity of the DAG is not especially efficient in the size of the DAG, and RB circuits contain both a large number of gates and barriers. As a short term fix, this PR checks whether or not two barriers are adjacent before performing the expensive "mergability" check.

```
$ asv profile 'randomized_benchmarking.RandomizedBenchmarkingBenchmark.time_ibmq_backend_transpile_single_thread\(\[\[0, 1\]\]\)' master
· Discovering benchmarks
·· Uninstalling from virtualenv-py3.5..
·· Building 04cc7a12 <master> for virtualenv-py3.5.......
·· Installing 04cc7a12 <master> into virtualenv-py3.5..
· Profile data does not already exist. Running profiler now.
·· Benchmarking virtualenv-py3.5
··· Running (randomized_benchmarking.RandomizedBenchmarkingBenchmark.time_ibmq_backend_transpile_single_thread--).
··· randomized_benchmarking.RandomizedBenchmarkingBenchmark.time_ibmq_backend_transpile_single_thread                                                                                                                             ok
··· =============== ============
       rb_pattern
    --------------- ------------
         [[0]]          n/a
        [[0, 1]]     2.16±0.04m
     [[0, 2], [1]]      n/a
    =============== ============

···· For parameters: [[0, 1]]
     /Users/kevin.krsulichibm.com/q/qiskit/.asv/env/1353f64e2ec13a8f2767ee4aed611581/lib/python3.5/site-packages/qiskit/__init__.py:54: RuntimeWarning: Could not import the Aer provider from the qiskit-aer package. Install qiskit-aer or check your installation.
       RuntimeWarning)
     /Users/kevin.krsulichibm.com/q/qiskit/.asv/env/1353f64e2ec13a8f2767ee4aed611581/lib/python3.5/site-packages/qiskit/__init__.py:62: RuntimeWarning: Could not import the IBMQ provider from the qiskit-ibmq-provider package. Install qiskit-ibmq-provider or check your installation.
       RuntimeWarning)
     Making the n=2 Clifford Table


Fri Nov 15 10:30:57 2019    /var/folders/yp/p2nbz2kx70s7dp_7n9yr0b0c0000gn/T/tmpif32klld

         640467365 function calls (638864104 primitive calls) in 258.092 seconds

   Ordered by: cumulative time

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.000    0.000  258.092  258.092 {built-in method builtins.exec}
        1    0.000    0.000  258.092  258.092 /anaconda3/lib/python3.5/site-packages/asv/benchmark.py:540(method_caller)
        1    0.055    0.055  258.092  258.092 /Users/kevin.krsulichibm.com/q/qiskit/test/benchmarks/randomized_benchmarking.py:105(time_ibmq_backend_transpile_single_thread)
        1    0.000    0.000  258.037  258.037 /Users/kevin.krsulichibm.com/q/qiskit/.asv/env/1353f64e2ec13a8f2767ee4aed611581/lib/python3.5/site-packages/qiskit/compiler/transpile.py:30(transpile)
        1    0.000    0.000  257.977  257.977 /Users/kevin.krsulichibm.com/q/qiskit/.asv/env/1353f64e2ec13a8f2767ee4aed611581/lib/python3.5/site-packages/qiskit/tools/parallel.py:67(parallel_map)
       50    0.002    0.000  257.976    5.160 /Users/kevin.krsulichibm.com/q/qiskit/.asv/env/1353f64e2ec13a8f2767ee4aed611581/lib/python3.5/site-packages/qiskit/compiler/transpile.py:209(_transpile_circuit)
       50    0.001    0.000  257.974    5.159 /Users/kevin.krsulichibm.com/q/qiskit/.asv/env/1353f64e2ec13a8f2767ee4aed611581/lib/python3.5/site-packages/qiskit/transpiler/transpile_circuit.py:25(transpile_circuit)
       50    0.001    0.000  257.872    5.157 /Users/kevin.krsulichibm.com/q/qiskit/.asv/env/1353f64e2ec13a8f2767ee4aed611581/lib/python3.5/site-packages/qiskit/transpiler/passmanager.py:182(run)
       50    0.048    0.001  257.871    5.157 /Users/kevin.krsulichibm.com/q/qiskit/.asv/env/1353f64e2ec13a8f2767ee4aed611581/lib/python3.5/site-packages/qiskit/transpiler/passmanager.py:222(_run_single_circuit)
       50    0.007    0.000  257.818    5.156 /Users/kevin.krsulichibm.com/q/qiskit/.asv/env/1353f64e2ec13a8f2767ee4aed611581/lib/python3.5/site-packages/qiskit/transpiler/runningpassmanager.py:123(run)
     1200    0.004    0.000  248.848    0.207 /Users/kevin.krsulichibm.com/q/qiskit/.asv/env/1353f64e2ec13a8f2767ee4aed611581/lib/python3.5/site-packages/qiskit/transpiler/runningpassmanager.py:145(_do_pass)
     1200    0.032    0.000  248.835    0.207 /Users/kevin.krsulichibm.com/q/qiskit/.asv/env/1353f64e2ec13a8f2767ee4aed611581/lib/python3.5/site-packages/qiskit/transpiler/runningpassmanager.py:172(_run_this_pass)
       50    0.014    0.000  196.733    3.935 /Users/kevin.krsulichibm.com/q/qiskit/.asv/env/1353f64e2ec13a8f2767ee4aed611581/lib/python3.5/site-packages/qiskit/transpiler/passes/barrier_before_final_measurements.py:32(run)
       50    0.165    0.003  193.466    3.869 /Users/kevin.krsulichibm.com/q/qiskit/.asv/env/1353f64e2ec13a8f2767ee4aed611581/lib/python3.5/site-packages/qiskit/transpiler/passes/merge_adjacent_barriers.py:50(run)
       50    0.244    0.005  184.345    3.687 /Users/kevin.krsulichibm.com/q/qiskit/.asv/env/1353f64e2ec13a8f2767ee4aed611581/lib/python3.5/site-packages/qiskit/transpiler/passes/merge_adjacent_barriers.py:87(_collect_potential_merges)
    19900    0.110    0.000  161.410    0.008 /Users/kevin.krsulichibm.com/q/qiskit/.asv/env/1353f64e2ec13a8f2767ee4aed611581/lib/python3.5/site-packages/networkx/algorithms/shortest_paths/generic.py:178(shortest_path_length)
    20600    6.215    0.000  161.273    0.008 /Users/kevin.krsulichibm.com/q/qiskit/.asv/env/1353f64e2ec13a8f2767ee4aed611581/lib/python3.5/site-packages/networkx/algorithms/shortest_paths/unweighted.py:25(single_source_shortest_path_length)
 13598261   27.507    0.000  147.814    0.000 /Users/kevin.krsulichibm.com/q/qiskit/.asv/env/1353f64e2ec13a8f2767ee4aed611581/lib/python3.5/site-packages/networkx/algorithms/shortest_paths/unweighted.py:69(_single_shortest_path_length)
     9950    0.066    0.000   89.778    0.009 /Users/kevin.krsulichibm.com/q/qiskit/.asv/env/1353f64e2ec13a8f2767ee4aed611581/lib/python3.5/site-packages/qiskit/dagcircuit/dagcircuit.py:1018(ancestors)
     9950    2.570    0.000   89.711    0.009 /Users/kevin.krsulichibm.com/q/qiskit/.asv/env/1353f64e2ec13a8f2767ee4aed611581/lib/python3.5/site-packages/networkx/algorithms/dag.py:75(ancestors)
     9950    0.065    0.000   86.992    0.009 /Users/kevin.krsulichibm.com/q/qiskit/.asv/env/1353f64e2ec13a8f2767ee4aed611581/lib/python3.5/site-packages/qiskit/dagcircuit/dagcircuit.py:1022(descendants)
     9950    2.514    0.000   86.926    0.009 /Users/kevin.krsulichibm.com/q/qiskit/.asv/env/1353f64e2ec13a8f2767ee4aed611581/lib/python3.5/site-packages/networkx/algorithms/dag.py:55(descendants)
 14420632   25.552    0.000   84.153    0.000 {method 'update' of 'dict' objects}
127615586   48.406    0.000   66.940    0.000 /Users/kevin.krsulichibm.com/q/qiskit/.asv/env/1353f64e2ec13a8f2767ee4aed611581/lib/python3.5/site-packages/qiskit/dagcircuit/dagnode.py:98(__hash__)
 16369456   13.397    0.000   25.537    0.000 /Users/kevin.krsulichibm.com/q/qiskit/.asv/env/1353f64e2ec13a8f2767ee4aed611581/lib/python3.5/site-packages/networkx/classes/coreviews.py:80(__getitem__)
   510991    0.446    0.000   25.174    0.000 /Users/kevin.krsulichibm.com/q/qiskit/.asv/env/1353f64e2ec13a8f2767ee4aed611581/lib/python3.5/site-packages/qiskit/dagcircuit/dagcircuit.py:730(<genexpr>)
   549543    3.253    0.000   24.492    0.000 /Users/kevin.krsulichibm.com/q/qiskit/.asv/env/1353f64e2ec13a8f2767ee4aed611581/lib/python3.5/site-packages/networkx/algorithms/dag.py:212(lexicographical_topological_sort)
 13832323   12.047    0.000   21.878    0.000 /Users/kevin.krsulichibm.com/q/qiskit/.asv/env/1353f64e2ec13a8f2767ee4aed611581/lib/python3.5/site-packages/networkx/classes/coreviews.py:101(__getitem__)
```

### CODE REVIEW

**Comment by kdk** [comment]
In 98df92f28 , I removed the a test from `test/python/transpiler/test_barrier_before_final_measurements.py`
```
-    def test_remove_barrier_in_different_qregs(self):
-        """Two measurements in different qregs to the same creg
-
-         q0:--|--[m]------     q0:---|--[m]------
-                  |                  |   |
-         q1:--|---|--[m]--  -> q1:---|---|--[m]--
-                  |   |                  |   |
-         c0:------.---|---     c0:-------.---|---
-            ----------.---        -----------.---
```

because this test covered `MergeAdjacentBarriers`, not `BarrierBeforeFinalMeasurements`. `BarrierBeforeFinalMeasurements` shouldn't internally call `MergeAdjacentBarriers`, rather this should be handled by the pass manager (this was discussed but left out of #2045 ). This removed test, for example, could be accomplished by running `MergeAdjacentBarriers` in a loop.

I'll open up a follow up issue to remove move `MergeAdjacentBarrier`s into the default pass managers and re-organize the tests.

File: test/python/transpiler/test_adjacent_barriers.py [hunk 1]
```diff
@@ -217,6 +217,62 @@ def test_barriers_with_blocking_obstacle_long(self):
...
+        """ Test that barriers don't merge if there is an obstacle that
+            is blocking
+
+                 ░       ░                     ░       ░
+        q_0: |0>─░───────░─           q_0: |0>─░───────░─
+                 ░       ░                     ░       ░
+        q_1: |0>─░───■─────    ->     q_1: |0>─░───■─────
+                 ░ ┌─┴─┐ ░                     ░ ┌─┴─┐ ░
+        q_2: |0>───┤ X ├─░─           q_2: |0>───┤ X ├─░─
+                   └───┘ ░                       └───┘ ░
```
Comment (ajavadia) [inline test/python/transpiler/test_adjacent_barriers.py]: for this test I'm not sure why the top-right barrier would not merge with the left barrier. What's blocking it?

Comment (kdk) [reply test/python/transpiler/test_adjacent_barriers.py]: Limitation of the drawer. There are only two barriers, left (on q0, q1) and right (q0, q2). Even though the barriers are adjacent (on q0) they can't be merged due to the dependency through the CX on q1,q2.

**Comment by kdk** [comment]
> Does this restore the original performance? I could not tell the "delta" from the performance numbers you posted.

Still needs to be re-benchmarked. Numbers above were from profiling, I'll post the performance delta when I get it.

**Comment by kdk** [comment]
From `asv continuous master merge-adjacent-barriers-first-check-adjacent --no-only-changed --bench randomized.*ibmq*`
```
       before           after         ratio
     [89cb7afc]       [98df92f2]
     <master>         <merge-adjacent-barriers-first-check-adjacent>
           failed           failed      n/a  randomized_benchmarking.RandomizedBenchmarkingBenchmark.time_ibmq_backend_transpile_single_thread([[0, 2], [1]])
       1.69±0.02m          1.39±0m    ~0.82  randomized_benchmarking.RandomizedBenchmarkingBenchmark.time_ibmq_backend_transpile([[0, 2], [1]])
          41.5±2s        28.9±0.6s    ~0.70  randomized_benchmarking.RandomizedBenchmarkingBenchmark.time_ibmq_backend_transpile([[0, 1]])
       2.06±0.01m       1.35±0.01m    ~0.66  randomized_benchmarking.RandomizedBenchmarkingBenchmark.time_ibmq_backend_transpile_single_thread([[0, 1]])
       40.1±0.02s        25.5±0.3s    ~0.64  randomized_benchmarking.RandomizedBenchmarkingBenchmark.time_ibmq_backend_transpile_single_thread([[0]])
          13.6±0s        8.02±0.1s    ~0.59  randomized_benchmarking.RandomizedBenchmarkingBenchmark.time_ibmq_backend_transpile([[0]])
```

The regression was 3-4x, so this PR only partially resolves the regression. I've got an approach for covering the rest through a larger change to `MergeAdjacentBarriers` that I'll put in a separate PR.

File: test/python/transpiler/test_adjacent_barriers.py (hunk 1 above)
Comment (ajavadia) [reply test/python/transpiler/test_adjacent_barriers.py]: gotcha.

**Comment by ajavadia** [comment]
Cool, thanks for checking the numbers. 

I think this optimization here is good to do in general, but I'm confused about why #3273 was needed in the first place. Why would we want to place barriers before measurements if there are no swappers invoked? The barrier insertion pass was created explicitly because the swappers sometimes brought some measurements to the middle of the circuit, and we wanted to avoid that for the sake of current hardware. In the absence of that use case, I'm not sure why we would want to do that (especially if there's a performance issue).

**Comment by kdk** [comment]
> Cool, thanks for checking the numbers.
> 
> I think this optimization here is good to do in general, but I'm confused about why #3273 was needed in the first place. Why would we want to place barriers before measurements if there are no swappers invoked? The barrier insertion pass was created explicitly because the swappers sometimes brought some measurements to the middle of the circuit, and we wanted to avoid that for the sake of current hardware. In the absence of that use case, I'm not sure why we would want to do that (especially if there's a performance issue).

From https://github.com/Qiskit/qiskit-terra/pull/3273#issuecomment-542857013 and the few following comments, the motivation was for consistency of output qobj. I agree that the current performance hit is a problem, and I'll revert that PR for the release and we can come up with a more performant approach.

This PR will still be beneficial for performance in cases where the mapper did run.

**Pull request merged**